packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,328 @@
+
+    function assertKeyDoesNotExist(key) {
+      assert(
+        !backingMap.has(key),
+        details`${q(keyName)} already registered: ${key}`,
+      );
+    }
+
+    function assertKeyExists(key) {
+      assert(backingMap.has(key), details`${q(keyName)} not found: ${key}`);
+    }
+
+    function virtualObjectKey(key) {
+      const instanceKey = valToSlotTable.get(key);
+      if (!instanceKey) {
+        return undefined;
+      } else {
+        const { type, virtual } = parseVatSlot(instanceKey);
+        if (type === 'object' && virtual) {
+          return `ws${storeID}.${instanceKey}`;
+        } else {
+          return undefined;
+        }
+      }
+    }
+
+    return harden({
+      has(key) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          return !!syscall.vatstoreGet(vkey);
+        } else {
+          return backingMap.has(key);
+        }
+      },
+      init(key, value) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          assert(
+            !syscall.vatstoreGet(vkey),
+            details`${q(keyName)} already registered: ${key}`,
+          );
+          syscall.vatstoreSet(vkey, JSON.stringify(m.serialize(value)));
+        } else {
+          assertKeyDoesNotExist(key);
+          backingMap.set(key, value);
+        }
+      },
+      get(key) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          const rawValue = syscall.vatstoreGet(vkey);
+          assert(rawValue, details`${q(keyName)} not found: ${key}`);
+          return m.unserialize(JSON.parse(rawValue));
+        } else {
+          assertKeyExists(key);
+          return backingMap.get(key);
+        }
+      },
+      set(key, value) {
+        const vkey = virtualObjectKey(key);
+        if (vkey) {
+          assert(
+            syscall.vatstoreGet(vkey),
+            details`${q(keyName)} not found: ${key}`,
+          );
+          syscall.vatstoreSet(vkey, JSON.stringify(m.serialize(harden(value))));
+        } else {
+          assertKeyExists(vkey);
+          backingMap.set(key, value);
+        }
+      },
+    });
+  }
+
+  /**
+   * Make a new kind of virtual object.
+   *
+   * @param {*} instanceMaker  A function of the form `instanceMaker(state)` that
+   *    will return a representative instance wrapped around the given state.
+   *
+   * @returns {*} a maker function that can be called to manufacture new
+   *    instance of this kind of object.  The parameters of the maker function
+   *    are those of the `initialize` method implemented in the representative
+   *    produced by the `instanceMaker` parameter.
+   */
+  function makeKind(instanceMaker) {
+    const kindID = `${allocateExportID()}`;
+    let nextInstanceID = 1;
+
+    function makeRepresentative(innerSelf) {
+      function ensureState() {
+        if (!innerSelf.rawData) {
+          innerSelf = cache.lookup(innerSelf.instanceKey);
+        }
+      }
+
+      function wrapData() {
```

Comment:
I had previously considered that myself and done it the way I had because I like having a named boundary around a complicated operation, but having now tried it as you suggest I'm thinking the new formulation is clearer as it doesn't suggest the (former) `wrapData` is something that might also be done elsewhere.  It's also two lines shorter.